pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java

```diff
@@ -621,25 +631,108 @@ private CompletableFuture<Integer> closeServiceUnit(String serviceUnit) {
+                                                long startTime,
+                                                CompletableFuture<Void> completionFuture) {
+        CompletableFuture<List<NamespaceBundle>> updateFuture = new CompletableFuture<>();
+
+        pulsar.getNamespaceService().getSplitBoundary(bundle, null).thenAccept(splitBundlesPair -> {
+            // Split and updateNamespaceBundles. Update may fail because of concurrent write to Zookeeper.
+            if (splitBundlesPair == null) {
+                String msg = format("Bundle %s not found under namespace", serviceUnit);
+                updateFuture.completeExceptionally(new BrokerServiceException.ServiceUnitNotReadyException(msg));
+                return;
+            }
+            List<CompletableFuture<Void>> futures = new ArrayList<>();
+            ServiceUnitStateData next = new ServiceUnitStateData(Owned, data.broker());
+            NamespaceBundles targetNsBundle = splitBundlesPair.getLeft();
+            List<NamespaceBundle> splitBundles = splitBundlesPair.getRight();
+            List<NamespaceBundle> successPublishedBundles = new CopyOnWriteArrayList<>();
+            for (NamespaceBundle sBundle : splitBundles) {
+                futures.add(pubAsync(sBundle.toString(), next).thenAccept(__ -> {
+                    successPublishedBundles.add(sBundle);
+                }));
+            }
+            NamespaceName nsname = bundle.getNamespaceObject();
+            FutureUtil.waitForAll(futures)
+                    .thenCompose(__ -> namespaceService.updateNamespaceBundles(nsname, targetNsBundle))
+                    .thenCompose(__ -> namespaceService.updateNamespaceBundlesForPolicies(nsname, targetNsBundle))
+                    .thenRun(() -> {
+                        bundleFactory.invalidateBundleCache(bundle.getNamespaceObject());
+                        updateFuture.complete(splitBundles);
+                    }).exceptionally(e -> {
+                        // Clean the new bundle when has exception.
+                        List<CompletableFuture<Void>> futureList = new ArrayList<>();
+                        for (NamespaceBundle sBundle : successPublishedBundles) {
+                            futureList.add(tombstoneAsync(sBundle.toString()).thenAccept(__ -> {}));
+                        }
+                        FutureUtil.waitForAll(futureList)
+                                .whenComplete((__, ex) -> {
+                                    if (ex != null) {
+                                        log.warn("Clean new bundles failed,", ex);
+                                    }
+                                    updateFuture.completeExceptionally(e);
+                                });
+                        return null;
+                    });
+        }).exceptionally(e -> {
+            updateFuture.completeExceptionally(e);
+            return null;
+        });
+
+        updateFuture.thenAccept(r -> {
```

Comment:
is it intentional to use thenAccept? It looks like a problem to call an async method in `thenAccept`

Reply:
Yes, you can see when this method is completed, it will call `completionFuture.complete(null);`. Why is it a problem to call an async method in `thenAccept`?

Reply:
Yes, that actually work. Usually, we are using the thenCompose to interact with an async operation. So that we don't need to complete the `completionFuture` in many places.

Reply:
Because it has retry logic, so I have to use `completionFuture ` here.